core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -54,108 +55,108 @@ import shapeless.ops.hlist.Tupler
  *   Http.server.serve(foobar.toService)
  * }}}
  */
-trait Endpoint[A] { self =>
+abstract class Endpoint[F[_], A](implicit E: Effect[F]) { self =>
```

Comment:
I'm interested in hearing your reasoning for this approach vs. threading `Effect[F]` into `map`, `product`, etc?

I've seen many libraries doing the later. For example, both Monix's [Iterant](https://github.com/monix/monix/blob/master/monix-tail/shared/src/main/scala/monix/tail/Iterant.scala) and FS2's [Stream](https://github.com/functional-streams-for-scala/fs2/blob/series/1.0/core/shared/src/main/scala/fs2/Stream.scala) (and many more) just require an implicit effect for their combinators.

Reply:
I've done it for the sake of hack & try it quickly. Going to do it per combinator, yes